tools/coredump/gosym.go

```diff
@@ -0,0 +1,162 @@
+package main
+
+import (
+	"context"
+	"debug/elf"
+	"debug/gosym"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/peterbourgon/ff/v3/ffcli"
+	"go.opentelemetry.io/ebpf-profiler/libpf"
+	"go.opentelemetry.io/ebpf-profiler/reporter"
+	"go.opentelemetry.io/ebpf-profiler/tools/coredump/modulestore"
+)
+
+type gosymCmd struct {
+	store    *modulestore.Store
+	casePath string
+}
+
+func newGosymCmd(store *modulestore.Store) *ffcli.Command {
+	args := &gosymCmd{store: store}
+
+	set := flag.NewFlagSet("gosym", flag.ExitOnError)
+	set.StringVar(&args.casePath, "case", "", "Path of the test case to debug")
+
+	return &ffcli.Command{
+		Name:       "gosym",
+		Exec:       args.exec,
+		ShortUsage: "gosym",
+		ShortHelp:  "Symbolize go test case",
+		FlagSet:    set,
+	}
+}
+
+func (cmd *gosymCmd) exec(context.Context, []string) (err error) {
+	// Validate arguments.
+	if cmd.casePath == "" {
+		return errors.New("please specify `-case`")
+	}
+
+	var test *CoredumpTestCase
+	test, err = readTestCase(cmd.casePath)
+	if err != nil {
+		return fmt.Errorf("failed to read test case: %w", err)
+	}
+
+	module, addrs, err := goModuleAddrs(test)
+	if err != nil {
+		return fmt.Errorf("failed to find go module addresses: %w", err)
+	}
+
+	goBinary, err := cmd.store.OpenReadAt(module.Ref)
+	if err != nil {
+		return fmt.Errorf("failed to open module: %w", err)
+	}
+	defer goBinary.Close()
+
+	locs, err := goSymbolize(goBinary, addrs)
+	if err != nil {
+		return fmt.Errorf("failed to symbolize: %w", err)
+	}
+
+	for addr, frame := range locs {
+		for _, originFrame := range addrs[addr] {
+			*originFrame = formatSymbolizedFrame(frame, false) + " (" + *originFrame + ")"
+		}
+	}
+
+	return writeTestCaseJSON(os.Stdout, test)
+}
+
+// goModuleAddrs returns the go module and the addresses to symbolize for it
+// mapped to pointers to the frames in c that reference them.
+func goModuleAddrs(c *CoredumpTestCase) (*ModuleInfo, map[libpf.AddressOrLineno][]*string, error) {
+	type moduleAddrs struct {
+		module *ModuleInfo
+		addrs  map[libpf.AddressOrLineno][]*string
+	}
+
+	moduleNames := map[string]*moduleAddrs{}
+	for i, module := range c.Modules {
+		moduleName := filepath.Base(module.LocalPath)
+		if _, ok := moduleNames[moduleName]; ok {
+			return nil, nil, fmt.Errorf("ambiguous module name: %q", moduleName)
+		}
+		moduleNames[moduleName] = &moduleAddrs{
+			module: &c.Modules[i],
+			addrs:  map[libpf.AddressOrLineno][]*string{},
+		}
+	}
+
+	// maxAddrs is the module with the most addresses to symbolize. We use this
+	// as a heuristic to determine which module is the Go module we're
+	// interested in.
+	// TODO(fg) alternatively we could extract all modules and run some check on
+	// them to see if they are go binaries. But this is more complex, so the
+	// current heuristic should be good enough for now.
```

Comment:
No need to extract: could just move the `elf.NewFile` call here and check for presence of `.gopclntab` section. Note that not every Go binary has this section exported -- it can be stripped -- but in that case the symbolization code doesn't work anyway.

Reply:
Good idea! f9bc8ea6147e7568ed9faec601772025fde1f4ad

Reply:
The section cannot be stripped, but the section header that points to it can be. Go runtime doesn't need the section header. Depending on build flags it may not even be emitted, though I forgot which flags exactly would cause it. Section headers in general are purely informal debug info: they are not required for executables to work and can be removed with `strip --strip-section-headers someexe`. It's possible to locate gopclntab heuristically with a pattern search, but I think this is fine as-is and we shouldn't over-engineer it for a debug command here.

Reply:
Ack, that makes sense. I added a TODO comment to potentially enhance the code in the future: 794aa88